### PR TITLE
Polaris text field

### PR DESCRIPTION
**Pull request by tomnez**
### Overview

Implements `polaris-text-field`.

Text field [Polaris docs](https://polaris.shopify.com/components/forms/text-field)
Text field [Github component](https://github.com/Shopify/polaris/tree/master/src/components/TextField)

Notes:

- Label Action: The `labelAction` property was implemented with the `content`, `accessibilityLabel`, and `onAction` attributes. `external`, `id`, and `url` were omitted - My thinking was that the parent component could link to external URLs if needed (via `onAction`) rather than forcing users to use one specific way to link to external URLs since there are possible vulnerabilities that we don't need to make ourselves responsible for.

- Chatted with @andrewpye about whether to use Ember's `{{textarea}}` and `{{input}}` helpers or use actual HTML elements. Ember doesn't have built-in bindings for a lot of the attributes Polaris implements, so we went with the latter. There are some methods in `polaris-text-field.js` to bind the `value` to the value of the textarea/inputs. If this isn't ideal I can swap it for using the Ember helpers, we'd just lose out on some of the `on{event}` attributes.

### Demos
(FYI this is the dummy app so icons aren't rendering)

**with prefix and suffix**
prefix-suffix

**placeholder**
placeholder

**numeric type**
number

**multiline**
multiline

**with help text**
helptext

**error state**
error

**disabled state**
disabled

**with action label**
action label

**with `connectedLeft` and `connectedRight` attributes**
connected-left-right

### CODE REVIEW

File: addon/components/polaris-text-field.js [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+    let { value, disabled, readOnly, error, multiline, focus } = this.getProperties('value', 'disabled', 'readOnly', 'error', 'multiline', 'focus');
+
+    let valueClass = value ? 'Polaris-TextField--hasValue' : '';
+    let disabledClass = disabled ? 'Polaris-TextField--disabled' : '';
+    let readOnlyClass = readOnly ? 'Polaris-TextField--readOnly' : '';
+    let errorClass = error ? 'Polaris-TextField--error' : '';
+    let multilineClass = multiline ? 'Polaris-TextField--multiline' : '';
+    let focusClass = focus ? 'Polaris-TextField--focus' : '';
+
+    return `Polaris-TextField ${ valueClass } ${ disabledClass } ${ readOnlyClass } ${ errorClass } ${ multilineClass } ${ focusClass }`;
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: lol, this gives you one long chunk of trailing whitespace if none of those flags are truthy 😛 How do you feel about pushing the classes onto an array and returning `classes.join(' ')`?

File: addon/components/polaris-text-field.js [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+   * Callback when focus is removed
+   *
+   * @property onBlur
+   * @public
+   * @type {Function}
+   * @default noop
+   */
+  onBlur() {},
+
+  addValueListener() {
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: This function should come after CPs

File: addon/components/polaris-text-field.js [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+
+    let valueClass = value ? 'Polaris-TextField--hasValue' : '';
+    let disabledClass = disabled ? 'Polaris-TextField--disabled' : '';
+    let readOnlyClass = readOnly ? 'Polaris-TextField--readOnly' : '';
+    let errorClass = error ? 'Polaris-TextField--error' : '';
+    let multilineClass = multiline ? 'Polaris-TextField--multiline' : '';
+    let focusClass = focus ? 'Polaris-TextField--focus' : '';
+
+    return `Polaris-TextField ${ valueClass } ${ disabledClass } ${ readOnlyClass } ${ errorClass } ${ multilineClass } ${ focusClass }`;
+  }),
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Make all of these internal-only CPs `.readOnly()`

File: addon/components/polaris-text-field.js [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+    }
+
+    return describedBy.join(' ');
+  }),
+
+  ariaLabelledBy: computed('id', function() {
+    return `${ this.get('id') }Label`;
+  }),
+
+  ariaInvalid: computed('error', function() {
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: This can just be `bool('error').readOnly()`

File: addon/components/polaris-text-field.js [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+    return !!this.get('error');
+  }),
+
+  inputType: computed('type', function() {
+    let type = this.get('type');
+
+    return type === 'currency' ? 'text' : type;
+  }),
+
+  input: computed('multiline', function() {
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Are you using this anywhere?

File: addon/components/polaris-text-field.js [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+    let multiline = this.get('multiline');
+
+    return multiline ? 'textarea' : 'input';
+  }),
+
+  minimumLines: computed('multiline', function() {
+    let multiline = this.get('multiline');
+
+    return (typeof multiline === 'number') ? multiline : 1;
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Use `typeOf` from `@ember/utils`

File: addon/components/polaris-text-field.js [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+  }),
+
+  heightStyle: computed('height', 'multiline', function() {
+    let { height, multiline } = this.getProperties('height', 'multiline');
+
+    return (height && multiline) ? htmlSafe(`height: ${ height }px`) : '';
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Is this better being `null` than an empty string?

File: addon/components/polaris-text-field.js [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+    });
+  },
+
+  didReceiveAttrs() {
+    this._super(...arguments);
+
+    let { input, focused } = this.getProperties('input', 'focused');
+
+    if (input && focused) {
+      input.focus();
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Isn't `input` a string here (either `'input'` or `'textarea'`)? 😕

File: addon/components/polaris-text-field.js [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+    }
+  },
+
+  willDestroyElement() {
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: This and `destroy` can go using `ember-lifeline` as described in other comments 🙌

File: addon/components/polaris-text-field.js [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+        id,
+        onChange,
+        value,
+        step,
+        min,
+        max,
+      } = this.getProperties('id', 'onChange', 'value', 'step', 'min', 'max');
+
+      step = step || 1;
+      min = min || -Infinity;
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: This is buggy since if `min` is specified as `0` you'll end up with `-Infinity` here. Maybe use `isPresent` or `isNone` here instead.

File: addon/components/polaris-text-field.js [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+      max = max || Infinity;
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Ditto here re. `0` bug

File: addon/components/polaris-text-field.js [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+      let numericValue = value ? parseFloat(value) : 0;
+
+      if (isNaN(numericValue)) {
+        return;
+      }
+
+      // Making sure the new value has the same length of decimal places as the
+      // step / value has.
+      let decimalPlaces = Math.max(dpl(numericValue), dpl(step));
+      let newValue = Math.min(max, Math.max(numericValue + steps * step, min));
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: These Polaris guys are pretty good at what they do 💯

File: addon/components/polaris-text-field.js [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+
+      onChange(String(newValue.toFixed(decimalPlaces)), id);
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: I know this is duplicated from the React code, but doesn't `toFixed` return a string anyway? 😕

File: addon/components/polaris-text-field.js [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+    },
+
+    handleExpandingResize(height) {
+      this.set('height', height);
+    },
+
+    handleChange(e) {
+      this.get('onChange')(e, this.get('id'));
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Check for `onChange` being present before calling it

File: addon/components/polaris-text-field.js [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+    handleFocus() {
+      this.set('focus', true);
+    },
+
+    handleBlur() {
+      this.set('focus', false);
+    },
+
+    handleClick() {
+      let [ field ] = this.element.querySelectorAll('input, textarea');
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Same here re. fetching `field` as in other comments above

File: addon/components/polaris-text-field/resizer.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
+import Component from '@ember/component';
+import layout from '../../templates/components/polaris-text-field/resizer';
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: Move this below the other `@ember` imports so the Ember imports are grouped together

File: addon/components/polaris-text-field.js [hunk 17]
```diff
@@ -0,0 +1,520 @@
+import Component from '@ember/component';
+import layout from '../templates/components/polaris-text-field';
+import { createUniqueIDFactory } from '../utils/other';
+import { assert } from '@ember/debug';
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Group the `@ember` imports together pls

File: addon/components/polaris-text-field/resizer.js [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+const REPLACE_REGEX = /[\n&<>]/g;
+
+const ENTITIES_TO_REPLACE = {
+  '&': '&amp;',
+  '<': '&lt;',
+  '>': '&gt;',
+  '\n': '<br>',
+};
+
+const replaceEntity = (entity) => {
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: No need for this to be defined as an arrow function 🤷‍♂️

File: addon/components/polaris-text-field/resizer.js [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+  return ENTITIES_TO_REPLACE[entity] || entity;
+};
+
+export default Component.extend({
+  classNames: ['Polaris-TextField__Resizer'],
+
+  'aria-hidden': true,
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: Not keen on this dasherized prop name, especially as it's not bound to anything on the DOM. Make this `ariaHidden` pls 👍

EDIT: if this is _meant_ to be bound to the resizer element, you need to add an attribute binding (which can accept the `ariaHidden` name anyway)

File: addon/components/polaris-text-field/resizer.js [hunk 20]
```diff
@@ -0,0 +1,132 @@
+import Component from '@ember/component';
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: This component's pretty genius really 🙌

File: addon/components/polaris-text-field/resizer.js [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+   * @public
+   * @type {Function}
+   * @default noop
+   */
+  onHeightChange(/* height */) {},
+
+  finalContents: computed('contents', function() {
+    let contents = this.get('contents');
+
+    contents = contents
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: The linter doesn't complain about having the `?` and `:` on the next line? 😮

File: addon/components/polaris-text-field/resizer.js [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+
+    return htmlSafe(contents);
+  }),
+
+  contentsForMinimumLines: computed('minimumLines', function() {
+    let minimumLines = this.get('minimumLines');
+    let content = '';
+
+    for (let line = 0; line < minimumLines; line++) {
+      content += '<br>';
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: I know this is borrowed from the React code, but maybe use string interpolation here?

File: addon/components/polaris-text-field/resizer.js [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    return htmlSafe(content);
+  }),
+
+  handleHeightCheck() {
+    let [ contentNode, minimumLinesNode ] = this.element.querySelectorAll('.Polaris-TextField__DummyInput');
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: Smart 👍

File: addon/components/polaris-text-field/resizer.js [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+    let newHeight = Math.max(contentHeight, minimumHeight);
+    let { currentHeight, onHeightChange } = this.getProperties('currentHeight', 'onHeightChange');
+
+    if (newHeight !== currentHeight) {
+      onHeightChange(newHeight);
+    }
+  },
+
+  addResizeHandler() {
+    this.element.addEventListener('resize', this.handleHeightCheck.bind(this));
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: Use `ember-lifeline` for this event listener as described in other comments - removes the need to explicitly remove the event listener and faff around with the `bind` call 🙌

File: addon/templates/components/polaris-text-field.hbs [hunk 25]
```diff
@@ -0,0 +1,118 @@
+<!-- template-lint invalid-interactive=false -->
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field.hbs]: Can this go right above the line that causes the linter to get angry?

File: addon/templates/components/polaris-text-field.hbs [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+
+<div class="Polaris-Labelled__LabelWrapper">
+  <div class="Polaris-Label">
+    <label id="{{id}}Label" for={{id}} class="Polaris-Label__Text">{{label}}</label>
+  </div>
+  {{#if labelAction}}
+    {{polaris-button
+      plain=true
+      text=labelAction.content
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field.hbs]: We tend to use `text` instead of `content` owing to possible issues if someone passes an Ember proxy object (which has its own `content` property already)

File: addon/templates/components/polaris-text-field.hbs [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+      >
+    {{/if}}
+
+    {{#if suffix}}
+      <div class="Polaris-TextField__Suffix" id="{{id}}Suffix">
+        {{render-content suffix}}
+      </div>
+    {{/if}}
+
+    {{#if (and (eq type "number") (not disabled))}}
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field.hbs]: Where possible we try to avoid using non-standard helpers to reduce the dependency on other addons. In this case it should be easy to define a `shouldShowSpinner` CP and use that instead 👍

File: addon/templates/components/polaris-text-field/connected.hbs [hunk 28]
```diff
@@ -0,0 +1,21 @@
+{{#if (or left right)}}
+  <div class="Polaris-Connected">
+    {{#if left}}
+      <div class="Polaris-Connected__Item Polaris-Connected__Item--connection">
+        {{render-content left}}
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field/connected.hbs]: I think this should just be `{{component left}}` instead of `render-content`, since it doesn't make sense to render text here as the styling looks wack if you do. Also the React implementation explicitly takes a `React.ReactNode` for this property, which in our Ember port usually translates to a component. We can still accept a string though as the `component` helper will work with that too 👍

File: addon/templates/components/polaris-text-field/connected.hbs [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+
+    {{#if right}}
+      <div class="Polaris-Connected__Item Polaris-Connected__Item--connection">
+        {{render-content right}}
+      </div>
+    {{/if}}
+  </div>
+{{else}}
+  {{yield}}
+{{/if}}
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field/connected.hbs]: Can you move this out into a separate `polaris-connected` component please, as per the React library? If you feel like it you could even do it under a separate PR and rebase this branch onto it 😛

File: addon/templates/components/polaris-text-field/resizer.hbs [hunk 30]
```diff
@@ -0,0 +1,6 @@
+<div class="Polaris-TextField__DummyInput">
+  {{finalContents}}
+</div>
+<div class="Polaris-TextField__DummyInput">
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field/resizer.hbs]: This shouldn't be rendered unless there's a `minimumLines` set

File: addon/templates/components/polaris-text-field/spinner.hbs [hunk 31]
```diff
@@ -0,0 +1,26 @@
+<!-- template-lint invalid-interactive=false -->
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field/spinner.hbs]: Move this above the line that causes the linter error for explicitness

File: addon/templates/components/polaris-text-field/spinner.hbs [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+
+<div
+  class="Polaris-TextField__Spinner"
+  aria-hidden="true"
+  onclick={{action onClick}}
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field/spinner.hbs]: Does this work as expected? I'd imagine it'll pass an event object as the first param, is that something that we should avoid?

File: tests/integration/components/polaris-text-field-test.js [hunk 33]
```diff
@@ -0,0 +1,222 @@
...
+    assert.equal(findAll(textFieldMultilineSelector).length, 1, 'it renders a textarea');
+  });
+
+  test('it renders an actionable link when a `labelAction` hash is present', async function(assert) {
+    this.setProperties({
+      labelActionText: label,
+      labelActionCalled: false
+    });
+
+    this.set('labelAction', () => {
```
Comment (andrewpye) [inline tests/integration/components/polaris-text-field-test.js]: Any reason this isn't in the `setProperties` call above?

File: tests/integration/components/polaris-text-field-test.js [hunk 34]
```diff
@@ -0,0 +1,222 @@
...
+    assert.equal(prefixNode.textContent.trim(), prefix, 'prefix value is correct');
+
+    let suffixNode = find(textFieldSuffixSelector);
+    assert.ok(suffixNode, 'it renders the suffix element');
+    assert.equal(suffixNode.textContent.trim(), suffix, 'suffix value is correct');
+  });
+
+  test('it renders as a connected component when `connectedLeft` or `connectedRight` is present', async function(assert) {
+    this.setProperties({
+      left,
```
Comment (andrewpye) [inline tests/integration/components/polaris-text-field-test.js]: These should be components :wink:

File: addon/components/polaris-text-field.js [hunk 35]
```diff
@@ -0,0 +1,520 @@
+import Component from '@ember/component';
+import layout from '../templates/components/polaris-text-field';
+import { createUniqueIDFactory } from '../utils/other';
```
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/components/polaris-text-field.js]: I'm good with removing it if you are, I was just matching Polaris as closely as possible. `guidFor` will give us some crazy long IDs, whereas the above gives us `TextField1, TextField2`, etc.

File: addon/components/polaris-text-field.js [hunk 36]
```diff
@@ -0,0 +1,520 @@
...
+import { assert } from '@ember/debug';
+import { computed } from '@ember/object';
+import { htmlSafe } from '@ember/string';
+import { debounceTask, runDisposables } from 'ember-lifeline';
```
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Ah excellent I'll give that try

File: addon/components/polaris-text-field.js [hunk 37]
```diff
@@ -0,0 +1,520 @@
...
+
+    id = id || getUniqueID();
+
+    this.setProperties({
+      height: null,
+      focus: false,
+      id
+    });
+
+    next(this, ()=> {
```
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/components/polaris-text-field.js]: I think I was having some issues but let me try again.

File: addon/components/polaris-text-field.js [hunk 38]
```diff
@@ -0,0 +1,520 @@
...
+
+  /**
+   * Visually hide the label
+   *
+   * @property labelHidden
+   * @public
+   * @type {Boolean}
+   * @default null
+   */
+  labelHidden: null,
```
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/components/polaris-text-field.js]: 😞 ahh

File: addon/components/polaris-text-field.js [hunk 39]
```diff
@@ -0,0 +1,520 @@
...
+   * @default null
+   */
+  connectedRight: null,
+
+  /**
+   * An element connected to the left of the input
+   *
+   * @property connectedLeft
+   * @public
+   * @type {String|Component}
```
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/components/polaris-text-field.js]: I'm ok with this change. Polaris _does_ let you pass just a plain string and renders it the same way shown in my screenshots, but if you're ok with not having that functionality then I'm good with just passing Component types.

File: addon/templates/components/polaris-text-field.hbs [hunk 40]
```diff
@@ -0,0 +1,116 @@
...
+    onclick={{action "handleClick"}}
+  >
+
+    {{#if prefix}}
+      <div class="Polaris-TextField__Prefix" id="{{id}}Prefix">
+        {{render-content prefix}}
+      </div>
+    {{/if}}
+
+    {{#if multiline}}
```
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/templates/components/polaris-text-field.hbs]: These two have slightly different attributes though, (for example `input` can take `max` and `min` whereas `textarea` doesn't). Should we use the wrapper and just let browsers do their thing when it comes to attributes that don't belong?

File: addon/components/polaris-text-field.js [hunk 41]
```diff
@@ -0,0 +1,520 @@
...
+   *
+   * @property labelAction
+   * @public
+   * @type {Object}
+   * @default null
+   *
+   * Currently supports:
+   * { onAction, content, accessibilityLabel }
+   */
+  labelAction: null,
```
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Hm no I can move it, I must have added this later on in development.

File: addon/components/polaris-text-field.js [hunk 42]
```diff
@@ -0,0 +1,520 @@
...
+   * @property onBlur
+   * @public
+   * @type {Function}
+   * @default noop
+   */
+  onBlur() {},
+
+  addValueListener() {
+    let [ field ] = this.element.querySelectorAll('input, textarea');
+    let _this = this;
```
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/components/polaris-text-field.js]: The context is different when passing methods to event listeners as callbacks though. If I were to use `this` in my arrow function callback, the "this" context refers to the html node, and not the Ember object.

File: addon/components/polaris-text-field.js (hunk 1 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: You'd still get the `Polaris-TextField` class though 😄 

I'm good with `.join(' ' )`, I'll update it.

File: addon/components/polaris-text-field.js (hunk 5 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Ah no, that's an artifact from before I switched over to using regular HTML inputs. Good eyes 👀

File: addon/components/polaris-text-field.js (hunk 7 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Works either way, I can update it.

File: addon/components/polaris-text-field.js (hunk 8 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Ahh more artifacts - yes I need to update the way the input is grabbed here.

File: addon/components/polaris-text-field.js (hunk 13 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Hm yeah it would. Weird, I'll get rid of the `String` call.

File: addon/components/polaris-text-field.js (hunk 14 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Do we need to do that? `onChange` is defaulted to a no-op, so this should be safe to just call at any time.

File: addon/components/polaris-text-field/resizer.js (hunk 19 above)
Comment (tomnez) [reply addon/components/polaris-text-field/resizer.js]: This value doesn't seem to ever change so I don't think it requires binding, it just needs an initial value. I'll update to `ariaHidden`.

File: addon/components/polaris-text-field/resizer.js (hunk 21 above)
Comment (tomnez) [reply addon/components/polaris-text-field/resizer.js]: Nope! But I feel like the Andy linter does, so if you'd like I can shift things onto one line, although we risk having it look different than the React code and harder to pick out differences.

File: addon/templates/components/polaris-text-field.hbs (hunk 40 above)
[the remark this answers is not in the document]
Comment (andrewpye) [reply addon/templates/components/polaris-text-field.hbs]: True. You could do something like create a curried component definition and use that, something like this:

```hbs
{{#with (component "wrapper-element" name=name <put all the common attributes here> as |inputElement|}}
  {{#if multiline}}
     {{component inputElement tagName="textarea" ...}}
  {{else}}
     {{component inputElement tagName="input" max=max min=min ...}}
  {{/if}}
{{/with}}
```

Not sure if that's just clunky though, what do you reckon? Definitely DRYs up the code, but maybe at the expense of readability?

File: addon/components/polaris-text-field.js (hunk 42 above)
[the remark this answers is not in the document]
Comment (andrewpye) [reply addon/components/polaris-text-field.js]: The `ember-lifeline` `this.addEventListener` method should take care of this for you 😉

File: addon/templates/components/polaris-text-field/spinner.hbs (hunk 32 above)
Comment (tomnez) [reply addon/templates/components/polaris-text-field/spinner.hbs]: Seems harmless but we can remove `onclick` if you want. I got into the habit of doing this so the action would get called first, but in this case it doesn't really matter so I can update it if you want.

File: addon/components/polaris-text-field.js (hunk 13 above)
Comment (andrewpye) [reply addon/components/polaris-text-field.js]: I'm good with it staying there since it's come from the React code, up to you 🤷‍♂️

File: tests/integration/components/polaris-text-field-test.js (hunk 33 above)
Comment (tomnez) [reply tests/integration/components/polaris-text-field-test.js]: Ah no, added it later on and missed this. Will move.

File: addon/components/polaris-text-field.js (hunk 14 above)
Comment (andrewpye) [reply addon/components/polaris-text-field.js]: Ah true, OK fine 😄

File: tests/integration/components/polaris-text-field-test.js (hunk 34 above)
Comment (tomnez) [reply tests/integration/components/polaris-text-field-test.js]: Well _now_ they should be, after making your updates 😛

File: addon/components/polaris-text-field/resizer.js (hunk 19 above)
Comment (andrewpye) [reply addon/components/polaris-text-field/resizer.js]: Yeah, I wasn't thinking about the value changing, just that it's something I've done elsewhere to declare something like an `ariaHidden` property and apply it to the component's root element using `attributeBindings` 😉

File: addon/components/polaris-text-field/resizer.js (hunk 21 above)
Comment (andrewpye) [reply addon/components/polaris-text-field/resizer.js]: Nah I don't mind, just surprised is all 😛

File: addon/templates/components/polaris-text-field.hbs (hunk 25 above)
Comment (andrewpye) [reply addon/templates/components/polaris-text-field.hbs]: ESLint lets you specify to ignore a rule just on the next line - if `template-lint` will let you do that, I'd prefer it 👍

File: addon/templates/components/polaris-text-field/spinner.hbs (hunk 32 above)
Comment (andrewpye) [reply addon/templates/components/polaris-text-field/spinner.hbs]: Don't mind too strongly either way, as long as the extra event param isn't visible anywhere 👍

File: addon/templates/components/polaris-text-field.hbs (hunk 40 above)
[the remark this answers is not in the document]
Comment (tomnez) [reply addon/templates/components/polaris-text-field.hbs]: If we have to clean this up then I'm good with the above, it _is_ DRY but yeah might be weird to read. I was originally going for something similar (hence my leftover `input` computed property) but ended up scrapping it. This seems ok though 👍

File: tests/integration/components/polaris-text-field-test.js (hunk 34 above)
Comment (tomnez) [reply tests/integration/components/polaris-text-field-test.js]: Just leaving this comment as an update to the above comments, we decided to keep the left and right connected attributes to accept both String or Component types to match React.

**Comment by tomnez** [comment]
@andrewpye updated. Some things to note about the updates -

- Left the `connectedLeft` and `connectedRight` tests as-is since we decided to keep it the same as React and accept a String or Component.
- `attributeBindings` on the `wrapper-element` were throwing an error regarding adding `attributeBindings` to a tagless component, so I left as-is.
- Computing `fieldElement` property wasn't working because it tries to compute before the DOM is available, so rather than using a CP I have it being set inside the `didInsertElement` hook to ensure that the field is available.
- Holy crap `ember-lifeline` is awesome.

File: addon/templates/components/polaris-text-field/spinner.hbs (hunk 31 above)
Comment (andrewpye) [reply addon/templates/components/polaris-text-field/spinner.hbs]: OK cool, thanks for checking 🙂

**Comment by andrewpye** [comment]
@tomnez few comments before I go over the code again:

- if this is now using #149 to bring in `polaris-connected`, can you change the base for the PR to that branch instead of `master` please?

- re. broken attribute binding on `wrapper-element`, check out #150 which should fix it. If it's still useful to you then we can merge it and you can rebase to bring it in, otherwise I think it's worth having anyway to avoid being bitten by the same gotcha in the future 😉 

- you've got a couple of failing tests which need to be addressed 👍

**Comment by andrewpye** [comment]
also re. `fieldElement` as a computed property, sounds like you were just accessing it too early somehow (i.e. before the component had rendered). Probably something that could be sorted out using the run loop but no biggie 🤷‍♂️

File: addon/components/polaris-text-field.js [hunk 43]
```diff
@@ -392,79 +389,82 @@ export default Component.extend({
...
-    this._super(...arguments);
-
-    let { input, focused } = this.getProperties('input', 'focused');
+  checkFocus() {
+    let { fieldElement, focused } = this.getProperties('fieldElement', 'focused');
 
-    if (input && focused) {
-      input.focus();
+    if (fieldElement && focused) {
+      this.$(fieldElement).focus();
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Any reason for the JQuery wrapper here? I'd prefer to not use it unless we have to - the ideal would be to be able to remove it from the addon altogether at some point...

File: addon/components/polaris-text-field.js [hunk 44]
```diff
@@ -0,0 +1,520 @@
...
+const dpl = (num) => {
+  (num.toString().split('.')[1] || []).length;
+};
+
+/**
+ * Polaris text-field component.
+ * See https://polaris.shopify.com/components/forms/text-field
+ */
+export default Component.extend({
+  classNameBindings: ['labelHidden:Polaris-Labelled--hidden'],
```
[the remark this answers is not in the document]
Comment (andrewpye) [reply addon/components/polaris-text-field.js]: #148 is now merged down to master, so you should be able to do some rebasing and refactor this to use `polaris-labelled` 👍

File: addon/components/polaris-text-field.js [hunk 45]
```diff
@@ -392,79 +389,82 @@ export default Component.extend({
...
+  }).readOnly(),
+
+  shouldShowSpinner: computed('disabled', 'type', function() {
+    let { type, disabled } = this.getProperties('type', 'disabled');
+
+    return (typeOf(type) === 'number') && !disabled;
+  }).readOnly(),
+
+  addValueListener() {
+    this.addEventListener('keyup', () => {
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Can you check that adding an event handler using an arrow function works in terms of tearing the handler down? I'd think `ember-lifeline` is intelligent enough to handle it properly but worth checking given that this is a leaky pattern for regular adding/removing event handlers...

File: addon/components/polaris-text-field.js [hunk 46]
```diff
@@ -392,79 +389,82 @@ export default Component.extend({
...
-    if (input && focused) {
-      input.focus();
+    if (fieldElement && focused) {
+      this.$(fieldElement).focus();
     }
   },
 
-  willDestroyElement() {
-    this.removeValueListener();
+  didInsertElement() {
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: Call `this._super(...arguments)` here 😉

File: addon/components/polaris-text-field.js [hunk 47]
```diff
@@ -392,79 +389,82 @@ export default Component.extend({
...
+    this.addValueListener();
+    this.checkFocus();
   },
 
-  destroy() {
-    runDisposables(this);
-
+  didReceiveAttrs() {
     this._super(...arguments);
+    this.checkFocus();
```
Comment (andrewpye) [inline addon/components/polaris-text-field.js]: This might need to be scheduled to use the run loop, since I'm not sure the attributes will have updated by the time this runs (you're probably alright though)

File: addon/components/polaris-text-field/resizer.js [hunk 48]
```diff
@@ -13,14 +14,15 @@ const ENTITIES_TO_REPLACE = {
...
-};
+}
 
-export default Component.extend({
+export default Component.extend(ContextBoundTasksMixin, ContextBoundEventListenersMixin, {
   classNames: ['Polaris-TextField__Resizer'],
+  attributeBindings: ['ariaHidden:aria-hidden'],
 
-  'aria-hidden': true,
+  ariaHidden: "true",
```
Comment (andrewpye) [inline addon/components/polaris-text-field/resizer.js]: Single quotes in JS files pls 😉

File: addon/templates/components/polaris-text-field.hbs [hunk 49]
```diff
@@ -1,13 +1,11 @@
-<!-- template-lint invalid-interactive=false -->
-
 <div class="Polaris-Labelled__LabelWrapper">
```
Comment (andrewpye) [inline addon/templates/components/polaris-text-field.hbs]: Rebase onto `master` to pull in `polaris-labelled` and refactor all this, it should be much more straightforward then 😉

File: addon/components/polaris-text-field.js (hunk 43 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Ahh no I can use `fieldElement.focus()` 🤦‍♂️

File: addon/components/polaris-text-field.js (hunk 45 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: I think we're safe, the example from their docs shows:

```javascript
didInsertElement() {
    this._super(...arguments);
    addEventListener(this, window, 'scroll', (e) => {
      this.set('windowScrollOffset', e.clientY);
    });
  },
```

The only difference is that I'm using the Mixin to create mine which uses the same `addEventListener` method above.

File: addon/components/polaris-text-field.js (hunk 47 above)
Comment (tomnez) [reply addon/components/polaris-text-field.js]: Did a quick test, we good -

focus

File: addon/components/polaris-text-field/resizer.js [hunk 50]
```diff
@@ -0,0 +1,128 @@
...
+      return;
+    }
+
+    let contentHeight = contentNode.offsetHeight;
+    let minimumHeight = minimumLinesNode.offsetHeight;
+    let newHeight = Math.max(contentHeight, minimumHeight);
+    let { currentHeight, onHeightChange } = this.getProperties('currentHeight', 'onHeightChange');
+
+    if (newHeight !== currentHeight) {
+      next(this, () => {
```
[the remark this answers is not in the document]
Comment (andrewpye) [reply addon/components/polaris-text-field/resizer.js]: I'd imagine the flow is something like this:

- `resizer` calls `onHeightChange`
- parent `text-field` component handles `onHeightChange`, which updates the `height` property
- parent `text-field` component passes updated `height` property into `resizer` component
- `resizer` component's `didUpdateAttrs` hook ends up invoking `onHeightChange` again because `currentHeight` hasn't updated yet, taking us around the loop a second time...

So the way to handle this would be to stop `onHeightChange` running more than once (especially in a single run loop). For that reason, `scheduleOnce` is probably more appropriate than `next` I'd say...

**Comment by tomnez** [comment]
@andrewpye updated and added docs.

[Pretty rendered docs](https://github.com/smile-io/ember-polaris/blob/polaris-text-field/docs/text-field.md#text-field) here.
